### PR TITLE
[Issue 13019][es-sink] Support event-time-based index name in ES Sink

### PR DESCRIPTION
**Pull request by fantapsody**
Fixes https://github.com/apache/pulsar/issues/13019

### Motivation

As described in the original issue, it's a common request to write data to event-time-based indices in logs and metrics use cases, therefore it would be very helpful to have builtin support in the ES sink.

### Modifications

*Describe the modifications you've done.*

### Verifying this change

- [ ] Make sure that the change passes the CI checks.
This change added tests and can be verified as follows:
  - *Added test cases for index name formatter*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

Check the box below or label this PR directly (if you have committer privilege).

Need to update docs? 

- [ ] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [ ] `no-need-doc` 
  
  (Please explain why)
  
- [x] `doc` 
  
  (If this PR contains doc changes)

### CODE REVIEW

File: pulsar-io/elastic-search/src/main/java/org/apache/pulsar/io/elasticsearch/ElasticSearchConfig.java [hunk 1]
```diff
@@ -50,7 +50,11 @@
     @FieldDoc(
         required = false,
         defaultValue = "",
-        help = "The index name that the connector writes messages to, the default is the topic name"
+        help = "The index name that the connector writes messages to, the default is the topic name."
```
Comment (Anonymitaet) [inline pulsar-io/elastic-search/src/main/java/org/apache/pulsar/io/elasticsearch/ElasticSearchConfig.java:53]:
```suggestion
        help = "The index name to which the connector writes messages. The default value is the topic name."
```

File: pulsar-io/elastic-search/src/main/java/org/apache/pulsar/io/elasticsearch/IndexNameFormatter.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.pulsar.client.api.schema.GenericObject;
+import org.apache.pulsar.functions.api.Record;
+
+/**
+ * A class that helps to generate the time-based index names.
+ *
+ * The formatter will find the pattern %{+<date-format>} in the format string,
```
Comment (Anonymitaet) [inline pulsar-io/elastic-search/src/main/java/org/apache/pulsar/io/elasticsearch/IndexNameFormatter.java:36]:
```suggestion
 * The formatter finds the pattern %{+<date-format>} in the format string
```
Write in the simple present tense as much as possible if you are covering facts that were, are, and forever shall be true. https://docs.google.com/document/d/1lc5j4RtuLIzlEYCBo97AC8-U_3Erzs_lxpkDuseU0n4/edit#bookmark=id.e8uqh1awkcnp

**Comment by Anonymitaet** [comment]
Discussed w/ @fantapsody 

1. @fantapsody  will add docs to [Elasticsearch sink connector - configuration table ](https://pulsar.apache.org/docs/en/next/io-elasticsearch-sink/#configuration) 
image

2. @fantapsody suggests generating configuration tables automatically rather than updating them manually. I've discussed w/ @urfreespace, found it is workable, and recorded an issue [here](https://github.com/apache/pulsar/issues/14393).

**Comment by fantapsody** [comment]
@eolivelli @freeznet @nlu90 @tuteng Could you please take a look?

**Review by freeznet — APPROVED** [review]

**Review by tuteng — APPROVED** [review]

File: site2/website-next/docs/io-elasticsearch-sink.md [hunk 3]
```diff
@@ -50,7 +50,7 @@ The configuration of the Elasticsearch sink connector has the following properti
 | Name | Type|Required | Default | Description 
 |------|----------|----------|---------|-------------|
 | `elasticSearchUrl` | String| true |" " (empty string)| The URL of elastic search cluster to which the connector connects. |
-| `indexName` | String| true |" " (empty string)| The index name to which the connector writes messages. |
+| `indexName` | String| true |" " (empty string)| The index name to which the connector writes messages. The default value is the topic name. It accepts date formats in the name to support event time based index with the pattern `%{+<date-format>}`. For example, suppose the event time of the record is 1645182000000L, the indexName is `logs-%{+yyyy-MM-dd}`, then the formatted index name would be `logs-2022-02-18`. |
```
Comment (Anonymitaet) [inline site2/website-next/docs/io-elasticsearch-sink.md:53]: Hi @fantapsody the docs should be added to `site2/docs/io-elasticsearch-sink.md` as well.

Comment (fantapsody) [reply site2/website-next/docs/io-elasticsearch-sink.md:53]: PTAL https://github.com/apache/pulsar/pull/14517

**Comment by nicoloboschi** [comment]
@tuteng @fantapsody @codelipenghui this pull is not present in Pulsar 2.10. I believe it has not been cherry-picked to branch-2.10 at all
I don't see it in the 2.10 release notes so there's no confusion for users. https://pulsar.apache.org/release-notes/#2100

**Pull request merged**